src/bunit.core/Rendering/TestRenderer.cs

```diff
@@ -117,6 +118,17 @@ public IReadOnlyList<IRenderedComponentBase<TComponent>> FindComponents<TCompone
 		where TComponent : IComponent
 		=> FindComponents<TComponent>(parentComponent, int.MaxValue);
 
+
+	/// <inheritdoc />
+	public void DetachComponents()
+	{
+		var rootComponents = renderedComponents.Values.OfType<IRenderedComponentBase<FragmentContainer>>().ToArray();
```

Comment:
What if we instead look for components of type `WrapperComponent`. That is the type of component that starts a new render tree when a call to Render or RenderComponent happens. Using that should also dispose any components added to `RenderTree`, which makes sense I think.

```suggestion
		var rootComponents = renderedComponents.Values.OfType<WrapperComponent>().ToArray();
```

Reply:
This is the relevant line where the WrapperComponent gets instantiated:

https://github.com/bUnit-dev/bUnit/blob/main/src/bunit.core/Rendering/TestRenderer.cs#L217

Reply:
I think it can be made to work by adding a `Detach` method to `WrapperComponent` that calls `renderHandle.Render(static builder => {})`.

Reply:
Then instead of calling `Render` in `DetachComponents` in `TestRenderer`, we call `Detach`. Thats also much cleaner code.

Reply:
I tried this out and we run into the same issues. Mainly because there is no `WrapperComponent` involved when `FindComponent` is called. Therefore the `cut` is not attached to any `WrapperComponent`. Only elements added by the `RenderTree.Add` method.

Reply:
I am sorry if I am rehashing this, but I am not sure I get why that would not work. If `WrapperComponent` triggers a render that does not have any child components, then ANY previously rendered components should be disposed, independent of they have been returned using `FindComponent` or not. 

Ill try out a few experiments myself to understand this better.

Reply:
Please feel free and don't worry. I like to discuss those things... even though via chat is sometimes a bit cumbersome 🗡️ 

>  If WrapperComponent triggers a render that does not have any child components, then ANY previously rendered components should be disposed [...]

That is true. Unfortunately there is no `WrapperComponent` involved anymore (due to `FindComponent`).

![image](https://user-images.githubusercontent.com/26365461/158361424-a6a22730-1f87-4539-859f-e6d44373c41f.png)

Reply:
Ahh yes, `renderedComponents` doesn't hold the raw components, but `IRenderedComponent`s. Can we use something else then `renderedComponents` to get at the rendered components.

In essence, we basically want the test renderer to dispose all components. A quick look through [Renderer.Dispose(bool)](https://source.dot.net/#Microsoft.AspNetCore.Components/RenderTree/Renderer.cs,982) makes it seems we can just call that in `DetachComponents`. Worth a try. Just need to make sure we can perform renders again after that.

Reply:
There is an inherit benefit of disposing the `Renderer` itself: It works with elements added via the `RootRenderTree`, but we have to create a new instance of the `TestRenderer` afterwards, which without further changes of the code will not be so easy todo. Especially because we are operating on the `Renderer` from the DI-Container. We have to account for multithreading, ... so at least we would need a lock inside `TestContext.DetachComponents`.

Furthermore I see a logical problem: We call `TestContext.DetachComponents` where the component under test might throw an exception. Now how do we test that?
`this.Renderer.UnhandledException` is not available anymore as we disposed the renderer and created a new one.

Also `WaitFor...` would suffer from similar issues.

Reply:
> Furthermore I see a logical problem: We call `TestContext.DetachComponents` where the component under test might throw an exception. Now how do we test that?

This is a good edge case, and we need a way to test this. Lets make sure it is possible to catch/verify that an exception is thrown when `DisposeComponents` is called.